### PR TITLE
Add Brotli splice cache slot handling

### PR DESCRIPTION
**Pull request by drinkbeer**
## Summary

Adds opt-in BrotliSplice slot handling for cached Brotli HTML responses, so a small per-request secret (e.g. a nonce or a `shopify_y`-style token) can be spliced into an otherwise fully-cached, pre-compressed Brotli body without recompressing it.

- Add BrotliSplice slot handling for cached Brotli HTML responses.
- Store splice slot metadata on cache writes and replace slot contents on cache hits.
- Add `ResponseBank::BrotliSpliceInjector` as the documented injector contract.
- Allow Rack apps to pass an injector builder directly to `ResponseBank::Middleware`.
- Allow Rails apps to configure the same builder with `config.response_bank.brotli_splice_injector`.
- Keep concrete metadata extraction in the application layer; ResponseBank owns only the generic Brotli splice hook and middleware wiring.
- Keep the Rack env injector key as an advanced compatibility hook.
- **`brotli_splice` is an _optional_ dependency.** It is a native C extension and only apps that opt into splice slots need it, so it is **not a runtime dependency** — consuming apps add `gem "brotli_splice"` to their own Gemfile and ResponseBank loads it lazily on first use (see _Review fixes_ below). It is declared as a **development dependency** in the gemspec so the gem can test the splice path itself; the CI matrix gemfiles inherit it through `gemspec` exactly as they do `minitest`/`rake`. The ordinary `brotli` compression gem stays a hard runtime dependency.
- Install Ubuntu `libbrotli-dev` in CI so the `brotli_splice` native extension can compile for the gem's own test suite.

## Usage shape

```ruby
use ResponseBank::Middleware, ->(env) { HtmlMetadataInjector.new(env) }
```

```ruby
config.response_bank.brotli_splice_injector =
  ->(env) { HtmlMetadataInjector.new(env) }
```

The consumer app owns `HtmlMetadataInjector` and may include `ResponseBank::BrotliSpliceInjector` to document the required methods. To enable the feature the consumer app must also add `gem "brotli_splice"` to its Gemfile.

## Review fixes

The second commit addresses review feedback and hardens edge cases:

- **README example correctness.** The documented injector reserved the wrong slot length and returned a replacement padded to the full slot width. Because BrotliSplice reserves the last 2 bytes of each slot as a fixed `\r\n` context suffix, the replacement never matched `replacement_length`, so `valid_replacement?` failed and every Brotli cache hit silently served the neutral placeholder. The example now reserves those 2 bytes and returns a replacement of exactly `replacement_length`, and documents that slot offset/length are byte offsets/counts (a character offset passes the bounds check but points at the wrong place once multi-byte UTF-8 precedes the placeholder).
- **Optional dependency + lazy load** (per review): in the gemspec `brotli_splice` moved from a runtime to a **development** dependency, so consumers no longer build the native extension, while the gem's own CI (the actionpack matrix gemfiles) still gets it through `gemspec`. At runtime it is loaded through a memoized `ensure_brotli_splice_loaded!` (`gem` + `require`, `rescue LoadError → warn + raise`) called at the top of the two methods that touch the native extension (`encode_body`, `replace_compressed_secret`). Apps without an injector never trigger the `require`; an opted-in app that forgot the gem fails loudly with a Gemfile hint. The loader deliberately sits **outside** the `rescue BrotliSplice::Error` block, so a missing gem surfaces as `LoadError` rather than a `NameError` from Ruby evaluating that (undefined) rescue constant.
- **Cover the miss → plain-body branch.** When a client accepts neither `br` nor `gzip`, `check_encoding` still defaults to `br`, so a cache miss with an injector produces Brotli metadata but must be served uncompressed via `replace_plain_body`. Added a middleware test (with a byte-offset injector) proving that branch is reached and splices the per-request value into the plain body — it is reachable, not dead code.
- **`valid_compressed_slot?` is now a pure predicate.** It previously coerced offset/length to integers and wrote them back into the slot hash; coercion moved to the caller, which passes locals to both the predicate and `BrotliSplice.replace` (behavior byte-for-byte identical).
- **`valid_html_slot?` rejects slots no larger than the 2-byte context suffix**, which have no replaceable region and would otherwise raise inside `BrotliSplice.encode`.
- **Observability:** log when `replace_compressed_secret` skips a slot (length mismatch / compressed slot out of bounds), mirroring the existing skip logs.
- Renamed `replace_compressed_body` → `replace_compressed_secret`; made `etag_matches?` use `start_with?` for the entity tag; aligned the README slot name to `shopify_y`; restored a trailing blank line in `gemfiles/actionpack_edge.gemfile`.

## Testing

- `ruby -c` passes for every changed Ruby file.
- Full suite, run in a Ruby 3.4 env with `libbrotli` + the `brotli_splice` extension available:

  ```bash
  bundle exec ruby -I. test/middleware_test.rb
  bundle exec rake test
  ```

  > The second commit adds `test_cache_miss_with_injector_serves_spliced_plain_body_to_client_without_br` plus the edge cases above, so run counts are higher than the original feature-commit run (previously middleware_test.rb: 19 runs / 101 assertions; full suite: 61 runs / 320 assertions). Final counts are confirmed by CI.

### CODE REVIEW

File: gemfiles/actionpack_7.2.gemfile [hunk 1]
```diff
@@ -3,4 +3,6 @@ source("https://rubygems.org")
 
 gemspec(path: '..')
 
+gem("brotli_splice", "= 0.1.1")
```
Comment (Edouard-chin) [inline gemfiles/actionpack_7.2.gemfile]: ✂️ We don't have to specific the gem in the Gemfile since it's already part of the gemspec. 

Those gemfiles are used to pin gems to a specific version so that we can test the library on different dependency versions and check whether we introduce new regression

File: lib/response_bank/brotli_splice_slot.rb [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+          body: prepared_body,
+          compressed_body: result[:data],
+          metadata: {
+            METADATA_KEY => {
+              'version' => METADATA_VERSION,
+              'slots' => [metadata_slot],
+            },
+          },
+        )
+      rescue BrotliSplice::Error, NoMethodError, ArgumentError => error
```
Comment (Edouard-chin) [inline lib/response_bank/brotli_splice_slot.rb]: Why do we need to rescue against NoMethodError ?

File: lib/response_bank/response_cache_handler.rb [hunk 3]
```diff
@@ -180,9 +186,9 @@ def etag_matches?(entity_tag, if_none_match)
 
       # strictly speaking an unquoted etag is not valid, yet common
       # to avoid unintended greedy matches in we check for naked entity then includes with quoted entity values
-      entity_tag = %{"#{entity_tag}"} unless entity_tag.starts_with?('"')
+      entity_tag = %{"#{entity_tag}"} unless entity_tag.start_with?('"')
```
Comment (Edouard-chin) [inline lib/response_bank/response_cache_handler.rb:189]: ✂️ This adds un-necessary diff hunks

File: lib/response_bank/brotli_splice_slot.rb [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  module BrotliSpliceSlot
+    INJECTOR_ENV_KEY = 'response_bank.html_metadata_injector'
+    METADATA_KEY = 'brotli_splice'
+    METADATA_VERSION = 1
+
+    EncodedBody = Struct.new(:body, :compressed_body, :metadata, keyword_init: true)
+
+    class << self
+      def encode_body(env, body, headers, compression_level:)
+        injector = env[INJECTOR_ENV_KEY]
```
Comment (Edouard-chin) [inline lib/response_bank/brotli_splice_slot.rb:17]: I'm not seeing any code that sets this environment, it only reads. Since it's not explained in the PR description, I assumed that application haves to inject an object into this env ?

If so, is this supposed to be a public API and should we document how to use it ?

**Review by Edouard-chin — COMMENTED** [review]
I think I miss the big pictures of how this new feature is supposed to be used which makes reviewing a bit tricky.

There is a lot of defensive code and early return statement, but I'm not sure when would this happen. E.g. why would an "injector" inside the `response_bank.html_metadata_injector` env not respond to `prepare_response_bank_brotli_splice`. Or why would that injector returns slots that aren't Integers etc...

It would be useful to add some description in the PR that explains what problem this feature solves, how application are supposed to use it

File: gemfiles/actionpack_7.2.gemfile (hunk 1 above)
Comment (drinkbeer) [reply gemfiles/actionpack_7.2.gemfile]: Removed the explicit `brotli_splice` entries from the ActionPack matrix Gemfiles. They now rely on the gemspec runtime dependency and only pin the Rails version under test.

File: lib/response_bank/response_cache_handler.rb (hunk 3 above)
Comment (drinkbeer) [reply lib/response_bank/response_cache_handler.rb:189]: Reverted this unrelated `starts_with?` / `start_with?` cleanup so the PR no longer carries this extra hunk.

File: lib/response_bank/brotli_splice_slot.rb (hunk 2 above)
Comment (drinkbeer) [reply lib/response_bank/brotli_splice_slot.rb]: Good call. The places that could have produced `NoMethodError` were malformed hook/metadata shapes, not an expected runtime condition. I replaced the broad `NoMethodError` rescues with explicit type guards for prepared payloads, slot metadata, bodies, and replacements. That preserves the fail-open behavior for malformed splice data, but no longer hides real bugs inside the injector implementation. Added a regression test for the malformed-slot fallback path.

File: lib/response_bank/brotli_splice_slot.rb (hunk 4 above)
Comment (drinkbeer) [reply lib/response_bank/brotli_splice_slot.rb:17]: Yes. The Storefront/SFR PR wires this up by inserting `HtmlMetadataInjectionMiddleware` before `ResponseBank::Middleware` and setting `env["response_bank.html_metadata_injector"] = injector` there. I agree PR #103 should still document the contract because this gem introduces the env key and method expectations, so I added a README section for the optional Rack env injector hook and updated the PR description.

**Comment by yoavweiss** [comment]
Comments from Claude:

1. The README example is inconsistent with the -2 context-suffix contract (will silently no-op at runtime). In the README, the placeholder is 36 bytes and length: PLACEHOLDER.bytesize (=36), but
 BrotliSplice.encode reserves the last 2 bytes as a fixed \r\n context suffix, so replacement_length comes back as 34. Then response_bank_brotli_splice_replacement does
 shopify_y.ljust(slot.fetch("replacement_length")) where shopify_y is a 36-byte UUID — ljust doesn't truncate, so it returns 36 bytes, valid_replacement? fails (36 ≠ 34), and the code silently serves the zeroed
 placeholder to the user. The test helper gets this right by adding a 2-byte ## suffix (length: placeholder.bytesize + PLACEHOLDER_SUFFIX.bytesize = 38 → replacement_length 36). The README should be corrected to
 teach the +2 accounting, because getting it wrong fails silently.
 
1. valid_compressed_slot? mutates its argument. A predicate named valid_..? writes back into the slot hash (slot['compressed_offset'] = offset). Surprising side effect in a validator; harmless today only because
 metadata is freshly deserialized per request. Split the coercion out from the validation.

 1. Single-slot encode vs. multi-slot replace. encode_body hard-rejects anything but exactly one slot (matching brotli_splice's "one slot per stream" limitation), but replace_compressed_body/replace_plain_body are
 written as reduce loops over N slots. It's dead generality that implies a capability the encoder doesn't have — either support multiple or drop the loop.

 1. Byte-offset vs. character-offset trap in the injector contract. The README injector computes offset = body.index(PLACEHOLDER) (a character index) and passes it to BrotliSplice.encode, which treats it as a byte
  offset. Any multi-byte UTF-8 before the placeholder will produce a wrong offset — and valid_html_slot? won't catch it (a char index is ≤ the byte length, so it passes). Since response_bank is now teaching this
 pattern, the docs should explicitly say offsets/lengths are bytes and show byteslice/.b usage.
 
  1. Minor / drive-by:
    - valid_html_slot? allows length > 0, but encode requires length > 2 (CTX_LEN). Tighten to > CTX_LEN so it fails cleanly instead of relying on the rescue.
    - The starts_with? → start_with? fix in etag_matches? is a legit but unrelated fix (and the sibling entity_tag.starts_with? two lines up is left as starts_with?). Fine to keep, but it's out of scope and could be
   its own commit.
    - Naming inconsistency: README uses name: "shopify-y", tests use "shopify_y".
    - The gemfiles/actionpack_edge.gemfile trailing-newline removal is noise.
    - Gemfile pins brotli_splice = 0.1.1 — a very early 0.x; worth confirming you're comfortable hard-depending on it.

 1. Test coverage gap. The miss→plain-body branch with metadata (body = [replace_plain_body(...)] if metadata) is effectively unreachable, because metadata only exists when content_encoding == 'br', which implies
 the client accepts br, which routes to the compressed branch. Either it's dead code (remove it) or there's a scenario I'm missing that should be tested. The genuinely exercised plain path is the cache-hit one in
 response_cache_handler, which is tested.

**Comment by drinkbeer** [comment]
Thanks for the thorough review — went through all six. Everything below is in the latest push (`7f073d3`).

**1. README example inconsistent with the −2 context-suffix contract (silent no-op).** Fixed — and thanks, that silent failure was the motivating bug. The example now sizes the slot to include the 2 reserved context-suffix bytes, and `response_bank_brotli_splice_replacement` returns a value guarded to `== replacement_length` instead of `ljust(replacement_length)` (which never truncates, so the 36-byte UUID never matched the 34-byte `replacement_length` and the splice was skipped). The prose that claimed the replacement must equal the full slot length is corrected too.

**2. `valid_compressed_slot?` mutates its argument.** Fixed. It's now a pure predicate; the integer coercion moved out into `replace_compressed_body`, which passes the coerced locals to both the predicate and `BrotliSplice.replace`. Behavior is byte-for-byte identical, with no more write-back into the slot hash.

**3. Single-slot encode vs. multi-slot replace.** Intentional — keeping it. `encode_body` enforces the current one-slot-per-stream limit (matching brotli_splice's own limitation), while the replace side is written as a fold over the stored `slots` metadata. Two reasons: (a) it keeps `replace_*` a pure function of the persisted metadata rather than baking in a hard-coded count, and (b) the per-slot chaining in `replace_compressed_body` is already the shape brotli_splice says multi-slot would require, so if/when the encoder relaxes to N slots only the encode-side guard changes. It's a 1-iteration fold today, not generality that can misbehave. If you'd prefer, I can add an explicit comment documenting the single-slot invariant rather than leaving the general shape unannotated.

**4. Byte-offset vs. character-offset trap.** Fixed in the docs. The README now states explicitly that slot `offset`/`length` are byte offsets/counts, locates the placeholder with `body.b.index(...)`, sizes it with `bytesize`, and calls out that a character offset silently passes the bounds check but points at the wrong byte once multi-byte UTF-8 precedes the placeholder.

**5. Minor / drive-by:**
- **`valid_html_slot?`** — tightened to `length > CONTEXT_SUFFIX_LENGTH` (2), so an unusable slot fails the validation gate cleanly instead of raising inside `encode` and landing in the rescue.
- **`start_with?`** — aligned the `entity_tag` line to `start_with?` as well (the sibling `if_none_match` line was already fixed, so this just finishes the pair). Agreed it's technically out of scope; I kept it since it's the same two-line block, but happy to split it into its own commit if you'd rather.
- **Naming** — README slot name changed to `"shopify_y"` to match the tests.
- **`gemfiles/actionpack_edge.gemfile`** — reverted; the PR no longer touches that file.
- **`brotli_splice = 0.1.1` pin** — intentional. It's a Shopify-owned gem and we're pinning exact during the initial rollout while the slot ABI (reserved 2-byte suffix, `replace` memcpy semantics) is still settling and the feature is flag-gated on the consumer side. We'll loosen the constraint once the format is stable.

**6. "miss → plain-body branch is unreachable."** It's actually reachable — the reasoning assumes `content_encoding == 'br'` implies the client accepts br, but `ResponseBank.check_encoding` *defaults* to `'br'` when `Accept-Encoding` contains neither `br` nor `gzip`:

```ruby
def check_encoding(env, default_encoding = 'br')
  if env['HTTP_ACCEPT_ENCODING'].to_s.include?('br')
    'br'
  elsif env['HTTP_ACCEPT_ENCODING'].to_s.include?('gzip')
    'gzip'
  else
    default_encoding # => 'br'
  end
end
```

So on a miss from a client sending `Accept-Encoding: identity` (or no header at all — common for crawlers / plain `curl`): `encode_body` runs, sets `metadata`, and caches the Brotli body; then at serve time the `include?('br')` check is false, so we fall to the `else`, drop `Content-Encoding`, and hit `replace_plain_body(...) if metadata`. You're right that the cache-*hit* plain path in `response_cache_handler` was the tested one — this is the sibling cache-*miss* path in the middleware, which was reachable but untested. Added a regression test that drives exactly it: `MiddlewareTest#test_cache_miss_with_injector_serves_spliced_plain_body_to_client_without_br`, using a byte-offset injector (the string-sub test double only matches once the Brotli body is decompressed). It asserts the cache stores the Brotli body + metadata, and that the served plain body has the per-request value spliced in with no `Content-Encoding`.

File: lib/response_bank/brotli_splice_slot.rb [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+              'slots' => [metadata_slot],
+            },
+          },
+        )
+      rescue BrotliSplice::Error, ArgumentError => error
+        ResponseBank.log("BrotliSplice encode skipped: #{error.class}")
+        nil
+      end
+
+      def replace_compressed_body(env, body, metadata)
```
Comment (yoavweiss) [inline lib/response_bank/brotli_splice_slot.rb]: `replace_compressed_secret`?

File: lib/response_bank/brotli_splice_slot.rb [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        injector = env[INJECTOR_ENV_KEY]
+        slots = metadata_slots(metadata)
+        return body unless injector && slots && body && body != ''
+
+        slots.reduce(body) do |current_body, slot|
```
Comment (yoavweiss) [inline lib/response_bank/brotli_splice_slot.rb]: Does this support multiple slots? Have we actually tested brotli_splice for multiple slots?

File: response_bank.gemspec [hunk 7]
```diff
@@ -21,6 +21,7 @@ Gem::Specification.new do |s|
 
   s.add_runtime_dependency("msgpack")
   s.add_runtime_dependency("brotli")
+  s.add_runtime_dependency("brotli_splice")
```
Comment (Edouard-chin) [inline response_bank.gemspec]: The usual pattern when a optional feature depends on a gem is to let the application add the optional gem in its own gemfile. This allow to not penalize applications that don't need this feature to not have to install and compile this dependency.

At runtime when the application opts-in for this feature, we usually do this:

```ruby
# lib/response_bank/brotli_splice_slot.rb

begin
  gem "brotli_splice"
  require "brotli_splice"
rescue LoadError => error
  warn "The Brotli splice splot feature requires "brotli_splice" gem. Please add it in your application gemfile"
  raise error
end
```

File: lib/response_bank/brotli_splice_slot.rb (hunk 6 above)
Comment (drinkbeer) [reply lib/response_bank/brotli_splice_slot.rb]: Currently one slot only. `encode_body` hard-rejects anything but exactly one slot (`return unless … slots.length == 1`), so the persisted metadata always carries a single slot — and `brotli_splice` itself only supports one spliceable slot per stream (per its README: *"Only one spliceable slot per stream (encode with multiple slots would require chaining encoders)"*).

And no, we haven't tested multiple slots — it isn't reachable through the supported path: the `== 1` guard means you can't produce multi-slot metadata via `encode_body`, and the gem has no multi-slot encode. The `slots.reduce` here is therefore a single-iteration fold in practice; it's written as a fold only so the replace side wouldn't need rewriting if `brotli_splice` ever gains multi-slot support and we relax the encode guard. Happy to collapse it to an explicit single-slot form, or add a comment documenting the invariant — whichever you'd prefer.

File: response_bank.gemspec (hunk 7 above)
Comment (drinkbeer) [reply response_bank.gemspec]: Adopted this exactly — thanks, much better fit than a hard runtime dep. In `5083475`:

- **Dropped the hard dependency**: removed `s.add_runtime_dependency("brotli_splice")` (this line) from the gemspec. `brotli` stays required — it's used for ordinary compression, not splicing.
- **Lazy load on opt-in**: added a memoized `ensure_brotli_splice_loaded!` with your `gem "brotli_splice"` / `require` / `rescue LoadError → warn + raise` pattern, and call it at the top of the two methods that actually touch native `BrotliSplice` — `encode_body` and `replace_compressed_secret`. `replace_plain_body` only delegates to the injector, so it doesn't need it. An app that never installs an injector never triggers the native `require`.

One subtlety worth flagging for the next reader: the `ensure_…!` call has to sit **outside** the `begin/rescue BrotliSplice::Error` block, not inside it. If the gem is missing and the guard were inside, Ruby evaluates the `BrotliSplice::Error` rescue class to test the match, that raises `NameError: uninitialized constant BrotliSplice`, and it masks the `LoadError` we actually want to surface. So each method is now `guards → ensure_brotli_splice_loaded! → begin BrotliSplice.* rescue BrotliSplice::Error, ArgumentError`.

Consuming-side follow-up: the app that opts in (SFR) needs `gem "brotli_splice", "= 0.1.1"` in **its own** Gemfile — I can't touch that repo from here. response_bank's own Gemfile keeps the gem so the test suite can exercise the splice path directly.

File: lib/response_bank/brotli_splice_slot.rb (hunk 6 above)
Comment (yoavweiss) [reply lib/response_bank/brotli_splice_slot.rb]: Adding a comment would help! thanks!!

**Review by yoavweiss — APPROVED** [review]
LGTM

**Review by Edouard-chin — APPROVED** [review]
